src/ClimaLSM.jl

```diff
@@ -1,329 +1,121 @@
+                                      )
+    args = (configuration, soil, vegetation)
+    return RootSoilModel{FT, typeof.(args)...}(args...)
 end
 
 function initialize_interactions(
-    land::LandModel{FT, RootSoilConfiguration{FT}}) where {FT}
+    land::RootSoilModel{FT, RootSoilConfiguration{FT}}) where {FT}
     soil_coords = land.soil.coordinates
     return (root_extraction = similar(soil_coords),)
 end
 
-function initialize(land::LandModel)
-    Y_soil, p_soil, coords_soil = initialize(land.soil)
-    Y_vegetation, p_vegetation, coords_vegetation = initialize(land.vegetation)
-    Y_river, p_river, coords_river = initialize(land.river)
-    Y_snow, p_snow, coords_snow = initialize(land.snow)
-    Y_carbon, p_carbon, coords_carbon = initialize(land.carbon)
+
+land_components(land::RootSoilModel) = (:soil, :vegetation)
+
+function initialize(land::AbstractLandModel)
+    components = land_components(land)
+    coords_list = map(components) do (component)
+        Domains.coordinates(getproperty(land, component))
+    end
+    Y_state_list = map(zip(components,coords_list)) do (component, coords)
+        getproperty(initialize_prognostic(getproperty(land, component), coords), component)
+    end
+    p_state_list = map(zip(components,coords_list)) do (component, coords)
+        getproperty(initialize_auxiliary(getproperty(land, component), coords), component)
+    end
     p_interactions  = initialize_interactions(land)
 
-    Y = ClimaCore.Fields.FieldVector(;
-                                     soil = Y_soil.soil,
-                                     vegetation = Y_vegetation.vegetation,
-                                     river = Y_river.river,
-                                     snow = Y_snow.snow,
-                                     carbon = Y_carbon.carbon
-    )
+    Y = ClimaCore.Fields.FieldVector(; NamedTuple(zip(components, Y_state_list))...)
     p = ClimaCore.Fields.FieldVector(;
-                                     p_interactions..., # for consistency, should this be p.interactions.root_extraction?
-                                     soil = p_soil.soil,
-                                     vegetation = p_vegetation.vegetation,
-                                     river = p_river.river,
-                                     snow = p_snow.snow,
-                                     carbon = p_carbon.carbon,
-    )
+                                     p_interactions...,
+                                     NamedTuple(zip(components, p_state_list))...)
     coords =
-        ClimaCore.Fields.FieldVector(; soil = coords_soil,
-                                     vegetation = coords_vegetation,
-                                     river = coords_river,
-                                     snow = coords_snow,
-                                     carbon = coords_carbon)
+        ClimaCore.Fields.FieldVector(; NamedTuple(zip(components, coords_list))...)
```

Comment:
Could create a helper function for this since it is used by `Y`, `p`, and `coords` in the same way.

Reply:
If it's OK with you, I'll wait on doing anything like this as `initialize` might change, and I'm also not sure a helper function would reduce the amount of code overall at this point..